src/progress.rs

```diff
@@ -141,99 +252,283 @@ impl ProgressSet {
     }
 
     /// Promote a learner to a peer.
     pub fn promote_learner(&mut self, id: u64) -> Result<(), Error> {
-        match self.progress.get_mut(&id) {
-            Some(progress) => if !progress.is_learner {
-                Err(Error::Exists(id, "voters"))?;
-            } else {
-                progress.is_learner = false;
-                self.configuration.voters.insert(id);
-                self.configuration.learners.remove(&id);
-            },
-            None => Err(Error::NotExists(id, "learners"))?,
+        debug!("Promote learner with id {}.", id);
+
+        if !self.configuration.learners.remove(&id) {
+            // Wasn't already a voter. We can't promote what doesn't exist.
+            return Err(Error::Exists(id, "learners"));
+        }
+        if !self.configuration.voters.insert(id) {
+            // Already existed, the caller should know this was a noop.
+            return Err(Error::Exists(id, "voters"));
+        }
+        if let Some(ref mut next) = self.next_configuration {
+            if !next.learners.remove(&id) {
+                // Wasn't already a voter. We can't promote what doesn't exist.
+                // Rollback change above.
+                self.configuration.voters.remove(&id);
+                self.configuration.learners.insert(id);
+                return Err(Error::Exists(id, "next learners"));
+            }
+            if !next.voters.insert(id) {
+                // Already existed, the caller should know this was a noop.
+                // Rollback change above.
+                self.configuration.voters.remove(&id);
+                self.configuration.learners.insert(id);
+                return Err(Error::Exists(id, "next voters"));
+            }
         }
+
         self.assert_progress_and_configuration_consistent();
         Ok(())
     }
 
     #[inline(always)]
     fn assert_progress_and_configuration_consistent(&self) {
         debug_assert!(
-            self.configuration
-                .voters
-                .union(&self.configuration.learners)
+            self.voter_ids()
+                .union(&self.learner_ids())
                 .all(|v| self.progress.contains_key(v))
         );
         debug_assert!(
             self.progress
                 .keys()
-                .all(|v| self.configuration.learners.contains(v)
-                    || self.configuration.voters.contains(v))
+                .all(|v| self.learner_ids().contains(v) || self.voter_ids().contains(v))
         );
         assert_eq!(
-            self.configuration.voters.len() + self.configuration.learners.len(),
+            self.voter_ids().len() + self.learner_ids().len(),
             self.progress.len()
         );
     }
+
+    /// Returns the minimal committed index for the cluster.
+    ///
+    /// Eg. If the matched indexes are [2,2,2,4,5], it will return 2.
+    pub fn minimum_committed_index(&self) -> u64 {
+        let mut matched = self
+            .voters()
```

Comment:
Is it correct when `progress.next_configuration` is not none? Suppose configuration contains voters `(1, 2, 3)`, and next_configuration contains voters `(2, 3, 4, 5)`, `voters` will be `(1, 2, 3, 4, 5)`, so the quorum is 3. However maybe we need acks from `(1, 2)` and `(3, 4, 5)` to commit a proposal.

Reply:
In this case the  quorum is 2 from the old, and 3 from the next, not 3. So we so acks from `(2,3,4)` is enough to commit, since both quorums are met.

I updated the code a bit (and I think fixed this that you were pointing it out).